### PR TITLE
Add new agentbeat with all beats shipped with Elastic Agent as a single beat

### PR DESCRIPTION
**Pull request by blakerouse**
## Proposed commit message

Add a new agentbeat that compiles into a single binary that provides subcommands for each contained beat. Contains auditbeat, filebeat, heartbeat, metricbeat, osquerybeat, and packetbeat.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~ **being this is just a combined beat, don't think we need to add a changelog for it (it just gets all the changes the other beats get)**

## Author's Checklist

- [x] auditbeat works 
- [x] filebeat works
- [x] heartbeat works
- [x] metricbeat works
- [x] osquerybeat works
- [x] packetbeat works

## How to test this PR locally

```
$ cd x-pack/agentbeat
$ mage package
$ cd build/distribution
$ tar zxvf agentbeat*.tar.gz
$ cd agentbeat*
$ ./agentbeat filebeat run -e
```

## Use cases

Provides a much smaller binary when all of the beats are combined into a single binary instead of each in there own binary.

### CODE REVIEW

**Review by belimawr — COMMENTED** [review]
That looks great @blakerouse !
Did you measure the disk savings? It would be nice to have it in the PR/Commit.

**Comment by leehinman** [comment]
Do any of the fields change when you ingest events with agentbeat?  I'm noticing some name changes in monitoring, input, and module names and I'm curious if that changes the emitted events in any way.

**Comment by blakerouse** [comment]
> That looks great @blakerouse ! Did you measure the disk savings? It would be nice to have it in the PR/Commit.

700M - combined total for auditbeat, filebeat, heartbeat, metricbeat, osquerybeat, packetbeat
207M - agentbeat (single build with auditbeat, filebeat, heartbeat, metricbeat, osquerybeat, packetbeat all in one)

493M in savings a reduction of 70% for just those beats.

**Note**: This was done on darwin/arm64. Each OS/arch will have a different result, but it should be in proportion.

**Comment by blakerouse** [comment]
> Do any of the fields change when you ingest events with agentbeat? I'm noticing some name changes in monitoring, input, and module names and I'm curious if that changes the emitted events in any way.

@leehinman That is a great question, I had to adjust the auditbeat system module name to `audit/system` because it collides with the same name of `system` with metricbeat. It only changes the registered name, I would hope that it doesn't affect the actual events being published. I don't know, I am no expert in how metricbeat works and if the registered module name affects the published events.

There is also an adjustment in the internal metrics reported for the beats, but its only for the name of the key for the running modules, and inputs. I don't believe this will have any affect, other than for debugging.

**Comment by amitkanfer** [comment]
@blakerouse - just to verify... cloudbeat is not included because it's in a different repo? there's no workaround for this?

**Comment by blakerouse** [comment]
@amitkanfer Correct. I would like to experiment once this lands a PR with cloudbeat merged in to see the cost to the agentbeat binary and the overall savings. Cloudbeat is only for Linux, so it would be a savings just for Linux. This PR currently provides savings for all platforms.

File: x-pack/auditbeat/module/system/system_name_agent.go
Comment (cmacknz) [inline x-pack/auditbeat/module/system/system_name_agent.go:11]: This breaks the `audit/system` input without another change.

In auditbeat modules are the input names. In the auditbeat spec file we have the same collision, so call this `audit/system` as you do here. However, because prior to this change the actual name of the module was `system` when we received the configuration from agent auditbeat stripped off the `audit` prefix.

https://github.com/elastic/beats/blob/4f46fcb888b7f12706d005b7e1d2d1dafcbe6fcd/x-pack/auditbeat/cmd/root.go#L31-L46

I think you have to stop doing this for the system module (or add the audit prefix to every auditbeat module). The other one in the auditbeat spec file is `audit/file_integrity` https://github.com/elastic/elastic-agent/blob/67ba0aa1869206af14c68a6ecd0a666c1ba3fc57/specs/auditbeat.spec.yml#L40

File: x-pack/osquerybeat/beater/osquerybeat.go
Comment (cmacknz) [inline x-pack/osquerybeat/beater/osquerybeat.go:132]: Is this a bug fix? Why was this necessary?

**Comment by cmacknz** [comment]
The output from the `/stats` endpoint is different under agentbeat than it is when running Filebeat directly. It appears we now include every monitoring namespace that was registered at init time for every beat by default, instead of only those for the binary we wanted to run.

When running `./agentbeat filebeat run -e` and then GET `http://localhost:5066/stats` I get the following which for example contains heartbeat stats. This data is collected by the monitoring Metricbeat, increasing the total data we would report for each Beat and including irrelevant details.

<details>

```json
{
  "amqp": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "beat": {
    "cpu": {
      "system": {
        "ticks": 20,
        "time": {
          "ms": 20
        }
      },
      "total": {
        "ticks": 62,
        "time": {
          "ms": 62
        },
        "value": 62
      },
      "user": {
        "ticks": 42,
        "time": {
          "ms": 42
        }
      }
    },
    "info": {
      "ephemeral_id": "30f48ca0-a7ac-4f0c-a0c3-f10070c360dc",
      "name": "filebeat",
      "uptime": {
        "ms": 31914
      },
      "version": "8.14.0"
    },
    "memstats": {
      "gc_next": 90639920,
      "memory_alloc": 68969232,
      "memory_sys": 107229452,
      "memory_total": 705409216,
      "rss": 186892288
    },
    "runtime": {
      "goroutines": 51
    }
  },
  "dhcpv4": {
    "parse_failures": 0,
    "total_packets": 0
  },
  "dns": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "filebeat": {
    "events": {
      "active": 0,
      "added": 63902,
      "done": 63902
    },
    "harvester": {
      "closed": 0,
      "open_files": 7,
      "running": 7,
      "skipped": 0,
      "started": 7
    },
    "input": {
      "log": {
        "files": {
          "renamed": 0,
          "truncated": 0
        }
      }
    }
  },
  "heartbeat": {
    "http": {
      "endpoint_starts": 0,
      "endpoint_stops": 0,
      "monitor_starts": 0,
      "monitor_stops": 0
    },
    "icmp": {
      "endpoint_starts": 0,
      "endpoint_stops": 0,
      "monitor_starts": 0,
      "monitor_stops": 0
    },
    "tcp": {
      "endpoint_starts": 0,
      "endpoint_stops": 0,
      "monitor_starts": 0,
      "monitor_stops": 0
    }
  },
  "http": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "icmp": {
    "duplicate_requests": 0,
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "libbeat": {
    "config": {
      "module": {
        "running": 0,
        "starts": 0,
        "stops": 0
      },
      "reloads": 1,
      "scans": 1
    },
    "output": {
      "batches": {
        "split": 0
      },
      "events": {
        "acked": 63902,
        "active": 0,
        "batches": 40,
        "dropped": 0,
        "duplicates": 0,
        "failed": 0,
        "toomany": 0,
        "total": 63902
      },
      "read": {
        "bytes": 0,
        "errors": 0
      },
      "type": "console",
      "write": {
        "bytes": 133231287,
        "errors": 0,
        "latency": {
          "histogram": {
            "count": 0,
            "max": 0,
            "mean": 0,
            "median": 0,
            "min": 0,
            "p75": 0,
            "p95": 0,
            "p99": 0,
            "p999": 0,
            "stddev": 0
          }
        }
      }
    },
    "pipeline": {
      "clients": 7,
      "events": {
        "active": 0,
        "dropped": 0,
        "failed": 0,
        "filtered": 0,
        "published": 63902,
        "retry": 0,
        "total": 63902
      },
      "queue": {
        "acked": 63902,
        "max_events": 3200
      }
    }
  },
  "memcache": {
    "unfinished_transactions": 0,
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "mongodb": {
    "unmatched_requests": 0
  },
  "mysql": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "nfs": {
    "unmatched_requests": 0
  },
  "pgsql": {
    "unmatched_responses": 0
  },
  "processor": {
    "add_host_metadata": {
      "fqdn_lookup_failed": 0
    }
  },
  "redis": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "registrar": {
    "states": {
      "cleanup": 0,
      "current": 0,
      "update": 0
    },
    "writes": {
      "fail": 0,
      "success": 0,
      "total": 0
    }
  },
  "system": {
    "cpu": {
      "cores": 10
    },
    "load": {
      "1": 2.8765,
      "15": 4.9624,
      "5": 3.8872,
      "norm": {
        "1": 0.2876,
        "15": 0.4962,
        "5": 0.3887
      }
    }
  },
  "thrift": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  }
}
```

</details>

The output from `http://localhost:5066/state` is similarly modified including both `filebeat_module` and `metricbeat_module` instead of just `module`. This one is less important because we can just stop collecting the `state` metricset for monitoring as it isn't actually used. https://github.com/elastic/elastic-agent/issues/4153

<details>

```json
{
  "beat": {
    "name": "Craigs-Macbook-Pro.local"
  },
  "filebeat_input": {
    "count": 0,
    "names": []
  },
  "filebeat_module": {
    "count": 0,
    "names": []
  },
  "heartbeat": {
    "endpoints": 0,
    "http": {
      "endpoints": 0,
      "monitors": 0
    },
    "icmp": {
      "endpoints": 0,
      "monitors": 0
    },
    "monitors": 0,
    "tcp": {
      "endpoints": 0,
      "monitors": 0
    }
  },
  "host": {
    "architecture": "arm64",
    "hostname": "Craigs-Macbook-Pro.local",
    "id": "48DA13D6-B83B-5C71-A4F3-494E674F9F37",
    "os": {
      "build": "23D60",
      "family": "darwin",
      "kernel": "23.3.0",
      "name": "macOS",
      "platform": "darwin",
      "version": "14.3.1"
    }
  },
  "management": {
    "enabled": false
  },
  "metricbeat_module": {
    "count": 0,
    "names": []
  },
  "output": {
    "batch_size": 0,
    "clients": 1,
    "name": "console"
  },
  "outputs": {
    "elasticsearch": {
      "cluster_uuid": ""
    }
  },
  "queue": {
    "name": "mem"
  },
  "service": {
    "id": "83cec81a-fe4a-4371-8341-881d668c1ebf",
    "name": "filebeat",
    "version": "8.14.0"
  }
}
```

</details>

File: x-pack/osquerybeat/beater/osquerybeat.go
Comment (aleksmaus) [reply x-pack/osquerybeat/beater/osquerybeat.go:132]: this was needed in osquerybeat in order to avoid having orphaned child osqueryd process
https://learn.microsoft.com/en-us/windows/win32/procthread/job-objects
"Operations performed on a job object affect all processes associated with the job object. Examples include enforcing limits such as working set size and process priority or terminating all processes associated with a job."

**Comment by blakerouse** [comment]
> The output from the `/stats` endpoint is different under agentbeat than it is when running Filebeat directly. It appears we now include every monitoring namespace that was registered at init time for every beat by default, instead of only those for the binary we wanted to run.
> 
> When running `./agentbeat filebeat run -e` and then GET `http://localhost:5066/stats` I get the following which for example contains heartbeat stats. This data is collected by the monitoring Metricbeat, increasing the total data we would report for each Beat and including irrelevant details.
> 
> The output from `http://localhost:5066/state` is similarly modified including both `filebeat_module` and `metricbeat_module` instead of just `module`. This one is less important because we can just stop collecting the `state` metricset for monitoring as it isn't actually used. [elastic/elastic-agent#4153](https://github.com/elastic/elastic-agent/issues/4153)

This PR was just my quick change to make it work. I wish that global `init()` was not used to register this stuff and it was only registered in the actual beater interface. If we could change to that model easily then we can make it where the contents are identical between each beat and the agentbeat.

File: x-pack/auditbeat/module/system/system_name_agent.go
Comment (blakerouse) [reply x-pack/auditbeat/module/system/system_name_agent.go:11]: Okay this will need to be conditional then between auditbeat and agentbeat. I will look at doing that, thanks for pointing it out.

 There was a lot of unique stuff per beat, that I learned doing this PR.

**Comment by cmacknz** [comment]
> This PR was just my quick change to make it work. I wish that global init() was not used to register this stuff and it was only registered in the actual beater interface. If we could change to that model easily then we can make it where the contents are identical between each beat and the agentbeat.

We can also account for the change in the processor configuration.

Besides the stats endpoint reporting extra data and the auditbeat system module name being changed in a breaking but fixable way, this looks like it should just work if we addressed those two points.

File: x-pack/auditbeat/module/system/system_name_agent.go
Comment (cmacknz) [reply x-pack/auditbeat/module/system/system_name_agent.go:11]: That `auditbeatCfg` only ever gets invoked when auditbeat is managed by agent.

When agent sends an auditbeat unit configuration, all the modules have an `audit/` prefix, which this code then strips out because auditbeat doesn't expect it.

In the system case we now include the `audit/` prefix, so we should stop stripping it out. It would actually be simpler if we kept the `audit/` prefix on all the modules when running under agentbeat, then we could remove the `strings.Split(rawIn.Type, "/")[1] ` in the code above and it should work.

I don't think you changed anything for Metricbeat but it is the same except that Metricbeat modules have a `/metrics` suffix to strip off instead of an `audit/` prefix. Auditbeat is based on Metricbeat where a module is the equivalent concept to an input in Filebeat.

**Comment by blakerouse** [comment]
I also believe this is wrong - https://github.com/elastic/beats/blob/main/x-pack/filebeat/cmd/root.go#L29

It needs to be changed to only call that transformation in the case that the beat is running.

**Comment by blakerouse** [comment]
I was able to get this in a state where their is no difference between a native beat versus a subcommand beat in agentbeat. Only a few modifications where required for that to work.

I have verified that a build of agentbeat works correctly with https://github.com/elastic/elastic-agent/pull/4516, for the integration testing suite.

**Comment by strawgate** [comment]
> > That looks great @blakerouse ! Did you measure the disk savings? It would be nice to have it in the PR/Commit.
> 
> 700M - combined total for auditbeat, filebeat, heartbeat, metricbeat, osquerybeat, packetbeat 207M - agentbeat (single build with auditbeat, filebeat, heartbeat, metricbeat, osquerybeat, packetbeat all in one)
> 
> 493M in savings a reduction of 70% for just those beats.
> 
> **Note**: This was done on darwin/arm64. Each OS/arch will have a different result, but it should be in proportion.

Can we make sure we also reference the total on-disk footprint when talking about these as these beats only account for 6 of the 9 binaries in the darwin package. Example: The Darwin aarch64 on-disk footprint is 1.05GB so a 493MB reduction is a 47% size reduction.

**Comment by cmacknz** [comment]
If I run filebeat through agentbeat with `./agentbeat filebeat run -e` with HTTP enabled in the `filebeat.yml` file in the same directory:

```yml
http.enabled: true
http.host: localhost
http.port: 5066
```

I still see global metrics from the other Beats present on the `/stats` endpoint, for example for filebeat I see heartbeat metrics.

```json
  "heartbeat": {
    "http": {
      "endpoint_starts": 0,
      "endpoint_stops": 0,
      "monitor_starts": 0,
      "monitor_stops": 0
    },
    "icmp": {
      "endpoint_starts": 0,
      "endpoint_stops": 0,
      "monitor_starts": 0,
      "monitor_stops": 0
    },
    "tcp": {
      "endpoint_starts": 0,
      "endpoint_stops": 0,
      "monitor_starts": 0,
      "monitor_stops": 0
    }
  },
```

<details>

```json
{
  "amqp": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "beat": {
    "cpu": {
      "system": {
        "ticks": 2,
        "time": {
          "ms": 2
        }
      },
      "total": {
        "ticks": 3,
        "time": {
          "ms": 3
        },
        "value": 3
      },
      "user": {
        "ticks": 1,
        "time": {
          "ms": 1
        }
      }
    },
    "info": {
      "ephemeral_id": "51894afd-8f09-483a-a35a-abca841f105a",
      "name": "filebeat",
      "uptime": {
        "ms": 117644
      },
      "version": "8.14.0"
    },
    "memstats": {
      "gc_next": 69289200,
      "memory_alloc": 52915008,
      "memory_sys": 70234524,
      "memory_total": 84422496,
      "rss": 143818752
    },
    "runtime": {
      "goroutines": 29
    }
  },
  "dhcpv4": {
    "parse_failures": 0,
    "total_packets": 0
  },
  "dns": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "filebeat": {
    "events": {
      "active": 0,
      "added": 0,
      "done": 0
    },
    "harvester": {
      "closed": 0,
      "open_files": 0,
      "running": 0,
      "skipped": 0,
      "started": 0
    },
    "input": {
      "log": {
        "files": {
          "renamed": 0,
          "truncated": 0
        }
      }
    }
  },
  "heartbeat": {
    "http": {
      "endpoint_starts": 0,
      "endpoint_stops": 0,
      "monitor_starts": 0,
      "monitor_stops": 0
    },
    "icmp": {
      "endpoint_starts": 0,
      "endpoint_stops": 0,
      "monitor_starts": 0,
      "monitor_stops": 0
    },
    "tcp": {
      "endpoint_starts": 0,
      "endpoint_stops": 0,
      "monitor_starts": 0,
      "monitor_stops": 0
    }
  },
  "http": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "icmp": {
    "duplicate_requests": 0,
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "libbeat": {
    "config": {
      "module": {
        "running": 0,
        "starts": 0,
        "stops": 0
      },
      "reloads": 1,
      "scans": 1
    },
    "output": {
      "batches": {
        "split": 0
      },
      "events": {
        "acked": 0,
        "active": 0,
        "batches": 0,
        "dropped": 0,
        "duplicates": 0,
        "failed": 0,
        "toomany": 0,
        "total": 0
      },
      "read": {
        "bytes": 0,
        "errors": 0
      },
      "type": "elasticsearch",
      "write": {
        "bytes": 0,
        "errors": 0,
        "latency": {
          "histogram": {
            "count": 0,
            "max": 0,
            "mean": 0,
            "median": 0,
            "min": 0,
            "p75": 0,
            "p95": 0,
            "p99": 0,
            "p999": 0,
            "stddev": 0
          }
        }
      }
    },
    "pipeline": {
      "clients": 0,
      "events": {
        "active": 0,
        "dropped": 0,
        "failed": 0,
        "filtered": 0,
        "published": 0,
        "retry": 0,
        "total": 0
      },
      "queue": {
        "acked": 0,
        "max_events": 3200
      }
    }
  },
  "memcache": {
    "unfinished_transactions": 0,
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "mongodb": {
    "unmatched_requests": 0
  },
  "mysql": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "nfs": {
    "unmatched_requests": 0
  },
  "pgsql": {
    "unmatched_responses": 0
  },
  "processor": {
    "add_host_metadata": {
      "fqdn_lookup_failed": 0
    }
  },
  "redis": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  },
  "registrar": {
    "states": {
      "cleanup": 0,
      "current": 0,
      "update": 0
    },
    "writes": {
      "fail": 0,
      "success": 0,
      "total": 0
    }
  },
  "system": {
    "cpu": {
      "cores": 10
    },
    "load": {
      "1": 5.98,
      "15": 6.6523,
      "5": 6.3247,
      "norm": {
        "1": 0.598,
        "15": 0.6652,
        "5": 0.6325
      }
    }
  },
  "thrift": {
    "unmatched_requests": 0,
    "unmatched_responses": 0
  }
}

```

</details>

I don't think this should block the PR, it isn't breaking. For agentbeat these metrics are collected by the agent's monitoring metricbeat and appear in the 30s metrics summaries in the agent logs for each beat. This will have the net effect of increasing the size of agent logs and metrics on disk and in Elasticsearch slightly if we don't do anything about it.

Is there any additional automated testing we should put in place for this change? The agent test framework only sanity checks basic operation of Filebeat and Metricbeat. It does not test auditbeat for example, although maybe this is covered by some of the tests in https://github.com/elastic/integrations/ and elsewhere.

**Review by cmacknz — APPROVED** [review]
Approving, we can work out any kinks later. Merging this doesn't switch agent to use agentbeat, it only makes agentbeat available from within Beats.

**Comment by blakerouse** [comment]
> If I run filebeat through agentbeat with `./agentbeat filebeat run -e` with HTTP enabled in the `filebeat.yml` file in the same directory:
> 
> ```yaml
> http.enabled: true
> http.host: localhost
> http.port: 5066
> ```
> 
> I still see global metrics from the other Beats present on the `/stats` endpoint, for example for filebeat I see heartbeat metrics.
> 
> ```json
>   "heartbeat": {
>     "http": {
>       "endpoint_starts": 0,
>       "endpoint_stops": 0,
>       "monitor_starts": 0,
>       "monitor_stops": 0
>     },
>     "icmp": {
>       "endpoint_starts": 0,
>       "endpoint_stops": 0,
>       "monitor_starts": 0,
>       "monitor_stops": 0
>     },
>     "tcp": {
>       "endpoint_starts": 0,
>       "endpoint_stops": 0,
>       "monitor_starts": 0,
>       "monitor_stops": 0
>     }
>   },
> ```
> 
> I don't think this should block the PR, it isn't breaking. For agentbeat these metrics are collected by the agent's monitoring metricbeat and appear in the 30s metrics summaries in the agent logs for each beat. This will have the net effect of increasing the size of agent logs and metrics on disk and in Elasticsearch slightly if we don't do anything about it.

I will look at reducing that in a follow up PR.

> 
> Is there any additional automated testing we should put in place for this change? The agent test framework only sanity checks basic operation of Filebeat and Metricbeat. It does not test auditbeat for example, although maybe this is covered by some of the tests in https://github.com/elastic/integrations/ and elsewhere.

I am working on a follow up PR already that runs the go and python integration framework with each beats tests using the agentbeat binary. That way we can ensure that it gets the same coverage as the individual beats.

File: x-pack/auditbeat/cmd/root.go
Comment (andrewkroh) [inline x-pack/auditbeat/cmd/root.go:57]: Without a length check this could panic. IIUC, the goal is to remove the `audit/` prefix, so I think using https://pkg.go.dev/strings#TrimPrefix would be more clear and safer (cannot panic based on user controlled inputs).

File: x-pack/auditbeat/module/system/system.go
Comment (andrewkroh) [inline x-pack/auditbeat/module/system/system.go:15]: I think this will affect `event.module` and possibly `event.dataset` field values (assuming this string goes into those fields). Within Fleet integrations these are `constant_keyword` fields with a static `value`. If the event document differs from the mapping then it will fail to index. So if this is true then we need to make more adjustments so that the field value does not change under agentbeat.

https://github.com/elastic/integrations/blob/77cf4288b00d76b0838f7e4275a297bcd37877c8/packages/auditd_manager/data_stream/auditd/fields/base-fields.yml#L10-L13

**Review by andrewkroh — COMMENTED** [review]
Have you reviewed the various `func init()` to see if any of them make assumptions about being isolated to a particular Beat? Here's one that comes to mind:

The `add_kubernetes_processor` has default indexers that are registered through `init` so the default indexers from every beat will be combined which would affect behavior. https://github.com/elastic/beats/blob/25786cdda70b31cb1738373265bf3a0f3dec76f6/packetbeat/processor/add_kubernetes_metadata/indexers.go#L30 https://github.com/elastic/beats/blob/29f0b4c805e2837dd1c2fa842b966f3182126f1b/filebeat/processor/add_kubernetes_metadata/matchers.go#L37

-------

Another side-effect of this will be that all beat processors will be available to every integration. This should not be a problem, but it is a change in behavior. For example, Filebeat specific processors will be available to Heartbeat inputs.

**Comment by strawgate** [comment]
Once we have an installer that includes Agentbeat can we benchmark the install time? It sounds like we're going to go from installing ~4000 files to installing <100 which will probably substantially reduce installation time

**Comment by blakerouse** [comment]
I am working on adjusting this PR where it only registers modules/processors when that sub-beat is being executed. That will ensure that they are registered the same as a single binary.

**Comment by pazone** [comment]
I see some errors when trying to package it locally:
```Error: running "go build -o build/golang-crossbuild/agentbeat-linux-arm64 -buildmode pie -trimpath -tags=agentbeat,withjournald -ldflags -s -X github.com/elastic/beats/v7/libbeat/version.buildTime=2024-04-11T12:07:32Z -X github.com/elastic/beats/v7/libbeat/version.commit=4d0b8b6ab6bc7207412467c3222ce2cdabef919b" failed with exit code 1
Error: failed building for linux/arm64: exit status 1
failed building for linux/arm64: exit status 1
Error: running "go build -o build/golang-crossbuild/agentbeat-linux-amd64 -buildmode pie -trimpath -tags=agentbeat,withjournald -ldflags -s -X github.com/elastic/beats/v7/libbeat/version.buildTime=2024-04-11T12:07:32Z -X github.com/elastic/beats/v7/libbeat/version.commit=4d0b8b6ab6bc7207412467c3222ce2cdabef919b" failed with exit code 1
Error: failed building for linux/amd64: exit status 1
failed building for linux/amd64: exit status 1
github.com/elastic/go-elasticsearch/v8/esapi: /usr/local/go/pkg/tool/linux_arm64/compile: signal: killed```

Should I worry about that?

**Comment by pazone** [comment]
Are we going to merge it soon? Or can use this branch to create the pipeline?

**Comment by blakerouse** [comment]
@andrewkroh I was able to adjust agentbeat and each of the beats to only register what each need. You will see the addition of `InitializeModule` and the automated way it is handled by mage to generate the code to call the initialization. This logic is only applied to areas where its critical that the original `func init()` is called when that specific beat is being used.

This change ensures that processors will not cross other beats, adjustments to kuberentes/docker processors will only take affect for each beat.

Another change is that `auditbeat` now uses its own registry to ensure that it doesn't collide with `metricbeat`. This allowed removing the changes that adjusted the name of the module and metricsets to ensure that when `agentbeat auditbeat` is called for the `system` module it used the `system` module for auditbeat, and when its `agentbeat metricbeat` its the `system` module for metricbeat.

**Comment by blakerouse** [comment]
@pazone I am waiting on a green CI and then I will get this merged so the PR for building can be created.

File: x-pack/auditbeat/cmd/root.go
Comment (blakerouse) [reply x-pack/auditbeat/cmd/root.go]: I was able to just revert this to how it was originally. So this change is no longer in this PR.

File: x-pack/auditbeat/module/system/system.go
Comment (blakerouse) [reply x-pack/auditbeat/module/system/system.go]: I have fixed this in the PR, it no longer changes the name of the module.

File: filebeat/autodiscover/builder/hints/logs.go
Comment (cmacknz) [inline filebeat/autodiscover/builder/hints/logs.go:40]: So we can't use `init()` anymore without side effects for agentbeat? 

I don't mind it, but this seems like something that will be forgotten or undone without some kind of automated enforcement.

Comment (blakerouse) [reply filebeat/autodiscover/builder/hints/logs.go:40]: Correct, caution will need to be taken in the case that a beat is modifying a behavior that is coming from libbeat. 

If the change can bleed into another beat then you need to use InitializeModule, otherwise init can be used. 

Like init is still being used in most places, but only the areas where it can cause an effect was it changed to InitializeModule.

Comment (blakerouse) [reply filebeat/autodiscover/builder/hints/logs.go:40]: It is also really hard to enforce with automation, because its conditional if it is required. It all depends on the context of the change.

It applied to anything that is modifying a behavior that is coming from libbeat or x-pack/libbeat.

**Comment by blakerouse** [comment]
All tests pass that are run by Jenkins, this is ready to be merged if okay with the implementation and change.

@pazone Is blocked on getting this merged to create the packaging job.

File: x-pack/auditbeat/module/system/login/login.go
Comment (andrewkroh) [inline x-pack/auditbeat/module/system/login/login.go:22]: I think all of these related module name changes should be reverted now that you have added a separate module registry.

File: auditbeat/cmd/show_linux.go
Comment (andrewkroh) [inline auditbeat/cmd/show_linux.go:66]: Why did show_linux.go need to be changed? Asking b/c I probably don't understand something. It was only adding itself to the `github.com/elastic/beats/v7/auditbeat/cmd.ShowCmd` which is isolated to the Auditbeat RootCmd. I don't see where this would cause a conflict under agentbeats combined command.

File: x-pack/auditbeat/module/system/login/login.go
Comment (blakerouse) [reply x-pack/auditbeat/module/system/login/login.go:22]: Really? This way is much cleaner using a single constant, instead of the same one in each module.

Not changing any behavior just more of a cleanup now. Seems like throwing away cleaner code.

File: auditbeat/cmd/show_linux.go
Comment (blakerouse) [reply auditbeat/cmd/show_linux.go:66]: Because it causes a circular import of cmd now, because the include had to be moved from main.go into root.go.

Without this change it will not compile. Even auditbeat will not compile without this change.

File: x-pack/auditbeat/module/system/login/login.go
Comment (andrewkroh) [reply x-pack/auditbeat/module/system/login/login.go:22]: True. I was more interested in reducing the changeset to only include the necessary changes to make it more digestible to reviewers.

File: auditbeat/cmd/show_linux.go
Comment (andrewkroh) [reply auditbeat/cmd/show_linux.go:66]: Oh, yes I see it now. That's very subtle.

**Review by andrewkroh — APPROVED** [review]
Assume it exists, can someone please point me at the buildkite job that periodically runs all Fleet integrations' system tests using the latest elastic-agent snapshot. I want to keep any on that as soon as this become available in a 8.14.0-SNAPSHOT.

**Comment by cmacknz** [comment]
Might be this one https://buildkite.com/elastic/elastic-package-test-with-integrations.

**Comment by cmacknz** [comment]
Merging this PR on its own isn't enough to update the agent build so you won't see any impact yet.

**Comment by cmacknz** [comment]
Merging so we can get the build pipeline and additional validation started.

**Comment by andrewkroh** [comment]
> Might be this one https://buildkite.com/elastic/elastic-package-test-with-integrations.

That pointed me in the right direction. I think it is this one. But we don't have any testing with the next snapshot.

https://buildkite.com/elastic/integrations-schedule-daily/builds/169

**Comment by blakerouse** [comment]
PR for packaging - https://github.com/elastic/beats/pull/38880

**Comment by blakerouse** [comment]
If your looking for an early build artifact - https://buildkite.com/elastic/beats-xpack-agentbeat-package/builds/17#018ed2c8-b857-429e-9d5c-f38f264fcc14

**Comment by lucabelluccini** [comment]
So this binary will fulfil the capabilities to implement the [inputs of integrations](https://www.elastic.co/guide/en/fleet/current/elastic-agent-inputs-list.html)?
Do we have coverage on each integration will be started correctly?

**Comment by cmacknz** [comment]
> So this binary will fulfil the capabilities to implement the [inputs of integrations](https://www.elastic.co/guide/en/fleet/current/elastic-agent-inputs-list.html)?

Yes, it is all of the Beats agent currently runs compiled into a single binary, with the individual beats accessible via sub-commands e.g `agentbeat filebeat run -e` should be equivalent to `filebeat run -e`

> Do we have coverage on each integration will be started correctly?

I am confident we don't have 100% test coverage for every feature of every integration, however the scope of what can go wrong with this change should be limited to interactions between the global state of each Beat. There are not many things in this category, most of them have been raised in this PR already. There are always unknown unknowns of course.

**Comment by andrewkroh** [comment]
PR to begin system testing with 8.14.0-SNAPSHOT for all integrations on [buildkite](https://buildkite.com/elastic/integrations-schedule-daily): https://github.com/elastic/integrations/pull/9585

**Pull request merged**
